Quark_Sketch/src/smc.cpp

```diff
@@ -0,0 +1,395 @@
+{
+	uart_irq_tx_enable(dev);
+
+	data_transmitted = false;
+	uart_fifo_fill(dev, (const uint8_t*)buf, len);
+	while (data_transmitted == false)
+		;
+
+	uart_irq_tx_disable(dev);
+}
+
+void start_arc(unsigned int reset_vector)
+{
+
+	if (reset_vector != 0) {
+		curie_shared_data->arc_start = reset_vector;
+	}
+
+	curie_shared_data->flags = 0;
+	for(int i = 0; i < 6400000; i++)
+	{
+	}
+
+	*SCSS_SS_CFG_REG |= ARC_RUN_REQ_A;
+
+	for(int i = 0; i < 6400000; i++)
+	{
+	}
+	
+	PRINT("ARC core started\n");
+}
+
+void reboot(void)
+{
+	*SCSS_SS_CFG_REG |= ARC_HALT_REQ_A;
+	*SCSS_RSTC = RSTC_WARM_RESET;
+}
+
+void main(void)
+{
+	// setup shared memory pointers for cdc-acm buffers
+	curie_shared_data->cdc_acm_buffers = &curie_shared_data->cdc_acm_buffers_obj;
+	curie_shared_data->cdc_acm_buffers_obj.rx_buffer = &curie_shared_data->cdc_acm_shared_rx_buffer;
+	curie_shared_data->cdc_acm_buffers_obj.tx_buffer = &curie_shared_data->cdc_acm_shared_tx_buffer;
+
+	softResetButton();
+	init_sharedMemory_com();
+	variantInit();
+
+	// start ARC core
+	uint32_t *reset_vector;
+	reset_vector = (uint32_t *)RESET_VECTOR;
+	start_arc(*reset_vector);
+	task_start(QUARK_SKETCH);
+	task_start(CDCACM_SETUP);
+	task_start(BAUDRATE_RESET);
+	task_start(USB_SERIAL);
+}
+
+extern "C" void cdcacm_setup(void)
+{
+	PRINT("cdcacm_setup task\r\n");
+	uint32_t baudrate, dtr = 0;
+	int ret;
+
+
+	dev = device_get_binding(CONFIG_CDC_ACM_PORT_NAME);
+
+	PRINT("Wait for DTR\n");
+	while (1) {
+		uart_line_ctrl_get(dev, LINE_CTRL_DTR, &dtr);
+		if (dtr)
+			break;
+		task_yield();
+	}
+	
+	PRINT("DTR set, start test\n");
+
+	/* They are optional, we use them to test the interrupt endpoint */
+	ret = uart_line_ctrl_set(dev, LINE_CTRL_DCD, 1);
+	if (ret)
+		PRINT("Failed to set DCD, ret code %d\n", ret);
+	
+	acm_tx_state = ACM_TX_READY;
+	acm_rx_state = ACM_RX_READY;
+	curie_shared_data->cdc_acm_buffers_obj.host_open = true;
+	
+	ret = uart_line_ctrl_set(dev, LINE_CTRL_DSR, 1);
+	if (ret)
+		PRINT("Failed to set DSR, ret code %d\n", ret);
+
+	/* Wait 1 sec for the host to do all settings */
+	sys_thread_busy_wait(1000000);
+
+	ret = uart_line_ctrl_get(dev, LINE_CTRL_BAUD_RATE, &baudrate);
+	if (ret)
+		PRINT("Failed to get baudrate, ret code %d\n", ret);
+	else
+		PRINT("Baudrate detected: %d\n", baudrate);
+
+	uart_irq_callback_set(dev, interrupt_handler);
+		
+	//reset head and tails values to 0
+	curie_shared_data->cdc_acm_shared_rx_buffer.head = 0;
+	curie_shared_data->cdc_acm_shared_rx_buffer.tail = 0;
+	curie_shared_data->cdc_acm_shared_tx_buffer.head = 0;
+	curie_shared_data->cdc_acm_shared_tx_buffer.tail = 0;
+	usbSetupDone = true;
+}
+
+extern "C" void baudrate_reset(void)
+{
+	PRINT("baudrate_reset task\r\n");
+	uint32_t baudrate, ret = 0;
+	while(!usbSetupDone)
+	{
+		task_yield();
+	}
+
+	ret = uart_line_ctrl_get(dev, LINE_CTRL_BAUD_RATE, &baudrate);	
+	
+	while(1)
+	{
+		ret = uart_line_ctrl_get(dev, LINE_CTRL_BAUD_RATE, &baudrate);
+		if(baudrate == 1200)
```

Comment:
Is 1200 the "magic" baud rate used for reset? can you put it in a #define please